File: src/main/java/com/nospawnn/ui/argeditors/IntEditor.java
package com.nospawnn.ui.argeditors;

import com.nospawnn.ui.listeners.ArgListener;

import javax.swing.*;
import java.awt.*;
import java.util.ArrayList;
import java.util.List;

public final class IntEditor extends ArgumentEditor {
    private JPanel contentPane;
    private JLabel label;
    private JSpinner editor;

    private final List<ArgListener> argListeners = new ArrayList<>();

    public IntEditor(String labelText, Integer initialValue) {
        this.label.setText(labelText);
        this.editor.setValue(initialValue);
    }

    @Override
    public JPanel getContentPane() {
        return contentPane;
    }

    @Override
    public void addListener(ArgListener al) {
        argListeners.add(al);
    }

    @Override
    public void notifyListeners() {
        for (var l : argListeners) l.argUpdated(this);
    }

    @Override
    public Integer getValue() {
        return (Integer) editor.getValue();
    }

    @Override
    public JLabel getLabel() {
        return this.label;
    }

    void createUIComponents() {
        this.label = new JLabel();
        this.editor = new JSpinner();
        this.editor.addChangeListener(e -> notifyListeners());
    }

    @Override
    public String getLabelText() {
        return this.label.getText();
    }

    @Override
    public Component getEditor() {
        return this.editor;
    }

}

File: src/main/kotlin/ui/argumenteditors/IntEditor.kt
package com.nospawnn.ui.argumenteditors

import com.nospawnn.ui.MainView
import javax.swing.JSpinner
import javax.swing.event.ChangeEvent
import javax.swing.event.ChangeListener

class IntEditor(argName: String, override val default: Int) : ArgumentEditor(argName) {
    override val editor = JSpinner().apply {
        value = default
    }

    init {
        editor.addChangeListener(object : ChangeListener {
            override fun stateChanged(e: ChangeEvent?) {
                (topLevelAncestor as MainView).updateOutput()
            }
        })
        addEditor(editor)
    }

    override fun getValue(): Int = editor.value as Int
}